File: web/lib/progression-api.ts
import { apiJson, send } from "./api";
import type {
  PatternReview,
  BlockParameters,
  UpdateBlockParametersRequest,
  ExerciseStrength,
  ProgressionLog,
} from "./types/progression";
import type { EnergyBalance } from "./types/plan";

// Server-only HTTP helpers for the progression console (IMPL-PROG-01). Do not
// import from client components — apiFetch reads server env + the Auth.js
// session. Mirrors the Android ProgressionApi contract exactly. All three
// endpoints are user-scoped with no path params (the backend derives the user
// from the auth token).

// ── Reads ────────────────────────────────────────────────────────────

// Per-movement-pattern weekly analysis (trend, fatigue, proposed set target).
export function getWeekReview(): Promise<PatternReview[]> {
  return apiJson<PatternReview[]>("/api/me/progression/week-review");
}

// The active training block's parameters (mode, rep ranges, RIR caps, …).
export function getBlockParameters(): Promise<BlockParameters> {
  return apiJson<BlockParameters>("/api/me/progression/block-parameters");
}

// Per-exercise estimated 1RM (the "weights"), heaviest-first.
export function getStrength(): Promise<ExerciseStrength[]> {
  return apiJson<ExerciseStrength[]>("/api/me/progression/strength");
}

// The engine's measured energy state (maintenance TDEE, mean intake, balance).
// Used by the plan-coherence view. May be absent on older backends — callers
// should tolerate a throw (the coherence page treats it as "no engine data").
export function getEnergyBalance(): Promise<EnergyBalance> {
  return apiJson<EnergyBalance>("/api/me/progression/energy-balance");
}

// The per-lift progression audit (IMPL-DELOAD-01 D4): one row per completed
// session, newest first — engine target vs performed top set + rationale.
export function getProgressionLog(exerciseId: string): Promise<ProgressionLog> {
  return apiJson<ProgressionLog>(
    `/api/me/progression/log?exerciseId=${encodeURIComponent(exerciseId)}`,
  );
}

// ── Mutations ────────────────────────────────────────────────────────

// Partial update of the block parameters; returns the updated block. Currently
// the console only edits `mode` (pinning it sets manualOverride server-side).
export function updateBlockParameters(
  body: UpdateBlockParametersRequest,
): Promise<BlockParameters> {
  return send<BlockParameters>(
    "/api/me/progression/block-parameters",
    "PUT",
    body,
  );
}

File: android/core-data/src/main/java/com/gte619n/healthfitness/data/workouts/progression/ProgressionApi.kt
package com.gte619n.healthfitness.data.workouts.progression

import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.PUT

/**
 * Retrofit service for the user-scoped progression console. All three endpoints
 * are user-scoped with no path params (the backend derives the user from the
 * auth token). Matches the backend contract exactly:
 *   GET api/me/progression/week-review        → [PatternReviewDto] array
 *   GET api/me/progression/block-parameters   → [BlockParametersDto]
 *   PUT api/me/progression/block-parameters   → [BlockParametersDto]
 */
interface ProgressionApi {
    @GET("api/me/progression/week-review")
    suspend fun weekReview(): List<PatternReviewDto>

    @GET("api/me/progression/block-parameters")
    suspend fun blockParameters(): BlockParametersDto

    @PUT("api/me/progression/block-parameters")
    suspend fun updateBlockParameters(@Body body: UpdateBlockParametersRequest): BlockParametersDto

    @GET("api/me/progression/strength")
    suspend fun strength(): List<ExerciseStrengthDto>

    @GET("api/me/progression/energy-balance")
    suspend fun energyBalance(): EnergyBalanceDto
}

/**
 * Partial update of the block parameters. Every field is nullable; an omitted
 * (null) field is left unchanged by the backend.
 */
data class UpdateBlockParametersRequest(
    val mode: String? = null,
    val successCriterion: String? = null,
    val expectedDriftPerDay: Double? = null,
)
